lib/plug/adapters/cowboy2.ex

```diff
@@ -0,0 +1,336 @@
+      Plug.Adapters.Cowboy2.shutdown MyPlug.HTTP
+
+  """
+  @spec http(module(), Keyword.t(), Keyword.t()) ::
+          {:ok, pid} | {:error, :eaddrinuse} | {:error, term}
+  def http(plug, opts, cowboy_options \\ []) do
+    run(:http, plug, opts, cowboy_options)
+  end
+
+  @doc """
+  Runs cowboy under https.
+
+  Besides the options described in the module documentation,
+  this module also accepts all options defined in [the `ssl`
+  erlang module] (http://www.erlang.org/doc/man/ssl.html),
+  like keyfile, certfile, cacertfile, dhfile and others.
+
+  The certificate files can be given as a relative path.
+  For such, the `:otp_app` option must also be given and
+  certificates will be looked from the priv directory of
+  the given application.
+
+  ## Example
+
+      # Starts a new interface
+      Plug.Adapters.Cowboy2.https MyPlug, [],
+        port: 443,
+        password: "SECRET",
+        otp_app: :my_app,
+        keyfile: "priv/ssl/key.pem",
+        certfile: "priv/ssl/cert.pem",
+        dhfile: "priv/ssl/dhparam.pem"
+
+      # The interface above can be shutdown with
+      Plug.Adapters.Cowboy2.shutdown MyPlug.HTTPS
+
+  """
+  @spec https(module(), Keyword.t(), Keyword.t()) ::
+          {:ok, pid} | {:error, :eaddrinuse} | {:error, term}
+  def https(plug, opts, cowboy_options \\ []) do
+    Application.ensure_all_started(:ssl)
+    run(:https, plug, opts, cowboy_options)
+  end
+
+  @doc """
+  Shutdowns the given reference.
+  """
+  def shutdown(ref) do
+    :cowboy.stop_listener(ref)
+  end
+
+  @doc """
+  A function for starting a Cowboy2 server under Elixir v1.5 supervisors.
+
+  It expects three options:
+
+    * `:scheme` - either `:http` or `:https`
+    * `:plug` - such as MyPlug or {MyPlug, plug_opts}
+    * `:options` - the server options as specified in the module documentation
+
+  ## Examples
+
+  Assuming your Plug module is named `MyApp` you can add it to your
+  supervision tree by using this function:
+
+      children = [
+        {Plug.Adapters.Cowboy2, scheme: :http, plug: MyApp, options: [port: 4040]}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  """
+  def child_spec(opts) do
+    scheme = Keyword.fetch!(opts, :scheme)
+    cowboy_opts = Keyword.fetch!(opts, :options)
+
+    {plug, plug_opts} =
+      case Keyword.fetch!(opts, :plug) do
+        {_, _} = tuple -> tuple
+        plug -> {plug, []}
+      end
+
+    {id, start, restart, shutdown, type, modules} =
+      child_spec(scheme, plug, plug_opts, cowboy_opts)
+
+    %{id: id, start: start, restart: restart, shutdown: shutdown, type: type, modules: modules}
+  end
+
+  defp child_spec(scheme, plug, opts, cowboy_options) do
+    [ref, trans_opts, proto_opts] = args(scheme, plug, opts, cowboy_options)
+
+    cowboy_function =
+      case scheme do
+        :http -> :start_clear
+        :https -> :start_tls
+      end
+
+    cowboy_args = [ref, trans_opts, proto_opts]
+
+    {
```

Comment:
Why are we building this tuple just to pattern match on it later on? Does it make sense to merge this into child_spec/1 at this point?